x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheFile.java

```diff
@@ -0,0 +1,293 @@
+        this.closed = false;
+        assert invariant();
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public long getLength() {
+        return tracker.getLength();
+    }
+
+    public Path getFile() {
+        return file;
+    }
+
+    @Nullable
+    public FileChannelRefCounted getChannelRefCounter() {
+        return channelRefCounter;
+    }
+
+    public boolean acquire(final EvictionListener listener) throws IOException {
+        ensureOpen();
+        boolean success = false;
+        if (refCounter.tryIncRef()) {
+            synchronized (this) {
+                try {
+                    ensureOpen();
+                    final Set<EvictionListener> newListeners = new HashSet<>(listeners);
+                    if (newListeners.add(Objects.requireNonNull(listener)) == false) {
+                        throw new IllegalStateException("Cannot add the same listener twice");
+                    }
+                    maybeOpenFileChannel(newListeners);
+                    listeners = Collections.unmodifiableSet(newListeners);
+                    success = true;
+                } finally {
+                    if (success == false) {
+                        refCounter.decRef();
+                    }
+                }
+            }
+        }
+        assert invariant();
+        return success;
+    }
+
+    public boolean release(final EvictionListener listener) {
+        boolean success = false;
+        synchronized (this) {
+            try {
+                final Set<EvictionListener> newListeners = new HashSet<>(listeners);
+                boolean removed = newListeners.remove(Objects.requireNonNull(listener));
+                maybeCloseFileChannel(newListeners);
+                listeners = Collections.unmodifiableSet(newListeners);
+                success = removed;
+            } finally {
+                if (success) {
+                    refCounter.decRef();
+                }
+            }
+        }
+        assert invariant();
+        return success;
+    }
+
+    private void closeInternal() {
+        assert Thread.holdsLock(this);
+        if (channelRefCounter != null) {
+            channelRefCounter.deleteAfterClose();
+        } else {
+            try {
+                Files.deleteIfExists(file);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        if (closed == false) {
+            final Set<EvictionListener> evictionListeners = new HashSet<>();
+            synchronized (this) {
+                if (closed == false) {
+                    closed = true;
+                    evictionListeners.addAll(listeners);
+                    refCounter.decRef();
+                }
+            }
+            if (evictionListeners != null) {
+                evictionListeners.forEach(listener -> listener.onEviction(this));
+            }
```

Comment:
I don't think we can? The `close()` method does not clean up the list of listeners but instead captures the registered listeners at closing time (end set the `closed` flag so that no new listeners can be registered). This is the duty of each listener to release and remove itself once it is notified of the eviction (and once the listener is done with any cache file resource)